src/Routing/Middleware/SecurityMiddleware.php

```diff
@@ -0,0 +1,197 @@
+    {
+        $this->checkValues($option, ['deny', 'sameorigin', 'allow-from']);
+
+        if ($option === 'allow-from') {
+            if (empty($url)) {
+                throw new InvalidArgumentException('The 2nd arg $url can not be empty when `allow-from` is used');
+            }
+            $option .= ' ' . $url;
+        }
+
+        $this->headers['x-frame-options'] = $option;
+
+        return $this;
+    }
+
+    /**
+     * X-XSS-Protection
+     *
+     * Reference: https://blogs.msdn.microsoft.com/ieinternals/2011/01/31/controlling-the-xss-filter
+     *
+     * Available Value: '1', '0', '1; mode=block'
+     *
+     * @param string $mode Mode value
+     * @return $this
+     */
+    public function setXssProtection($mode = 'block')
+    {
+        $mode = (string)$mode;
+
+        if ($mode === 'block') {
+            $mode = '1; mode=block';
+        }
+
+        $this->checkValues($mode, ['1', '0', '1; mode=block']);
+        $this->headers['x-permitted-cross-domain-policies'] = $mode;
+
+        return $this;
+    }
+
+    /**
+     * X-Permitted-Cross-Domain-Policies
+     *
+     * Reference: https://www.adobe.com/devnet/adobe-media-server/articles/cross-domain-xml-for-streaming.html
+     *
+     * Available Value: 'all', 'none', 'master-only', 'by-content-type', 'by-ftp-filename'
+     *
+     * @param string $policy Policy value
+     * @return $this
+     */
+    public function setCrossDomainPolicy($policy = 'all')
+    {
+        $this->checkValues($policy, ['all', 'none', 'master-only', 'by-content-type', 'by-ftp-filename']);
+        $this->headers['x-permitted-cross-domain-policies'] = $policy;
+
+        return $this;
+    }
+
+    /**
+     * Convenience method to check if a value is in the list of allowed args
+     *
+     * @throws \InvalidArgumentException Thrown when a value is invalid.
+     * @param string $value Value to check
+     * @param array $allowed List of allowed values
+     * @return void
+     */
+    protected function checkValues($value, array $allowed)
+    {
+        if (!in_array($value, $allowed)) {
+            throw new InvalidArgumentException(sprintf(
+                    'Invalid arg `%s`, use one of these: %s',
+                    $value,
+                    implode(', ', $allowed)
+                )
+            );
+        }
+    }
+
+    /**
+     * Serve assets if the path matches one.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+     * @param \Psr\Http\Message\ResponseInterface $response The response.
+     * @param callable $next Callback to invoke the next middleware.
+     * @return \Psr\Http\Message\ResponseInterface A response
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, $next)
+    {
+        foreach ($this->headers as $header => $value) {
+            $response = $response->withHeader($header, $value);
```

Comment:
I think you want to modify the response after calling $next.  Right now headers will be lost if the lower layers create a new instance.